### PR TITLE
CLOUDP-193004: Add command “atlas backups compliancepolicy policies update” L1 support

### PR DESCRIPTION
**Pull request by VincentPikand**
## Proposed changes

Added the `atlas backups compliancepolicy policies update` command. A few notable changes here:
- I refactored the store to be more granular and composable. The idea is that there is an interface for every way you can update BCP and then your command code picks the ones it needs, composing a store. This also moves away some of the updating logic from the commands.
- The store refactor is "backwards compatible" meaning the behaviour of the store for the existing commands is unchanged. That will be another PR.

### Command Output

```
Your backup compliance policy is being updated with the following policies:

ID                         FREQUENCY INTERVAL   FREQUENCY TYPE   RETENTION  
64b14832d20ba902974585e0   1                    daily            1 day    
64b320f01c8b0e2993f3b0a8   4                    hourly           1 week
64b9d2abdc2db9d07e7415a8   -                    on demand        5 days
```

_Jira ticket:_ CLOUDP-193004

## Checklist

- [ ] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added any necessary documentation in document requirements section listed in [CONTRIBUTING.md](https://github.com/mongodb/mongodb-atlas-cli/blob/master/CONTRIBUTING.md) (if appropriate)
- [ ] I have addressed the @mongodb/docs-cloud-team comments (if appropriate)
- [ ] I have updated [test/README.md](https://github.com/mongodb/mongodb-atlas-cli/blob/master/test/README.md) (if an e2e test has been added)
- [ ] I have run `make fmt` and formatted my code

### CODE REVIEW

File: internal/cli/atlas/backup/compliancepolicy/setup.go [hunk 1]
```diff
@@ -32,12 +32,11 @@ import (
...
-	fs          afero.Fs
-	path        string
-	confirm     bool
-	EnableWatch bool
+	policy  *atlasv2.DataProtectionSettings
+	store   store.CompliancePolicy
+	fs      afero.Fs
+	path    string
+	confirm bool
 }
```
Comment (VincentPikand) [inline internal/cli/atlas/backup/compliancepolicy/setup.go:40]: removed `EnableWatch` as it already exists in `cli.WatchOpts`

File: internal/cli/atlas/backup/compliancepolicy/policies/update/update.go [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+}
+
+To get the ID of a policy item, run "atlas backups compliancepolicy policies describe".
+
+Finally, run the command as such: "atlas backups compliancepolicy policies update --file /path/to/file"
+`
+
+var errorCode500Template = `received an internal error on the server side, but we would encourage you to double check your inputs.
+For this command, invalid inputs are known to cause internal errors in some situations`
+
```
Comment (VincentPikand) [inline internal/cli/atlas/backup/compliancepolicy/policies/update/update.go:86]: Not entirely sure if this should be included, but the message is true. If the idea is OK, I'm going to double check the wording of it with the docs team.

File: internal/cli/atlas/backup/compliancepolicy/policies/update/update.go [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+}
+
+func (opts *Opts) Run(policyItem *atlasv2.DiskBackupApiPolicyItem) error {
+	result, httpResponse, err := opts.store.UpdatePolicyItem(opts.ProjectID, policyItem)
+	if err != nil {
+		if httpResponse != nil && httpResponse.StatusCode == 500 {
+			return fmt.Errorf("%v: %w", errorCode500Template, err)
+		}
+		return err
+	}
```
Comment (VincentPikand) [inline internal/cli/atlas/backup/compliancepolicy/policies/update/update.go:117]: usage of the err500 template

File: test/e2e/atlas/backup_compliancepolicy_policies_update_test.go [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+		AuthorizedEmail:    atlasv2.PtrString(authorizedEmail),
+		OnDemandPolicyItem: &initialItem,
+		ProjectId:          atlasv2.PtrString(g.projectID),
+	}
+	path := "./policies_update_setup.json"
+	createJSONFile(t, compliancePolicy, path)
+	res, err := setupCompliancePolicy(g.projectID, path)
+	r.NoError(err)
+
+	t.Run("policies update atomic", func(t *testing.T) {
```
Comment (VincentPikand) [inline test/e2e/atlas/backup_compliancepolicy_policies_update_test.go:56]: There will be another scenario (namely the interactive run) with the L3 PR

File: internal/cli/atlas/backup/compliancepolicy/policies/update/update.go [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+	"github.com/spf13/afero"
+	"github.com/spf13/cobra"
+	atlasv2 "go.mongodb.org/atlas-sdk/v20230201004/admin"
+)
+
+type combinedStore interface {
+	store.CompliancePolicyItemUpdater
+	store.ProjectLister
+	store.CompliancePolicyDescriber
+}
```
Comment (gssbzn) [inline internal/cli/atlas/backup/compliancepolicy/policies/update/update.go:37]: see https://github.com/mongodb/mongodb-atlas-cli/pull/2180#discussion_r1302950959

File: internal/cli/atlas/backup/compliancepolicy/policies/update/update.go (hunk 3 above)
Comment (gssbzn) [inline internal/cli/atlas/backup/compliancepolicy/policies/update/update.go:117]:
```suggestion
	result, err := opts.store.UpdatePolicyItem(opts.ProjectID, policyItem)
	if err != nil {
		if atlasv2.IsErrorCode(err, 500) {
			return fmt.Errorf("%s: %w", errorCode500Template, err)
		}
		return err
	}
```

File: test/e2e/atlas/backup_compliancepolicy_policies_update_test.go (hunk 4 above)
Comment (gssbzn) [reply test/e2e/atlas/backup_compliancepolicy_policies_update_test.go:56]: will it be able to run with the same project?

File: internal/cli/atlas/backup/compliancepolicy/policies/update/update.go (hunk 5 above)
Comment (VincentPikand) [reply internal/cli/atlas/backup/compliancepolicy/policies/update/update.go:37]: fixed the same way that fixed in that PR

File: internal/cli/atlas/backup/compliancepolicy/policies/update/update.go [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+)
+
+var updateTemplate = `Your backup compliance policy is being updated with the following policies:
+
+ID	FREQUENCY INTERVAL	FREQUENCY TYPE	RETENTION
+{{- range .ScheduledPolicyItems}}
+{{.Id}}	{{if eq .FrequencyType "hourly"}}{{.FrequencyInterval}}{{else}}-{{end}}	{{.FrequencyType}}	{{.RetentionValue}} {{.RetentionUnit}}
+{{- end}}
+{{if .OnDemandPolicyItem}}{{.OnDemandPolicyItem.Id}}	-	{{.OnDemandPolicyItem.FrequencyType}}	{{.OnDemandPolicyItem.RetentionValue}} {{.OnDemandPolicyItem.RetentionUnit}}{{end}}
+`
```
Comment (gssbzn) [inline internal/cli/atlas/backup/compliancepolicy/policies/update/update.go:53]: from the other Pr let's keep these simple
```suggestion
var updateTemplate = `Your backup compliance policy is being updated
`
```

File: internal/cli/atlas/backup/compliancepolicy/policies/update/update.go [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+var updateWatchTemplate = `Your backup compliance policy has been updated with the following policies:
+
+ID	FREQUENCY INTERVAL	FREQUENCY TYPE	RETENTION
+{{- range .ScheduledPolicyItems}}
+{{.Id}}	{{if eq .FrequencyType "hourly"}}{{.FrequencyInterval}}{{else}}-{{end}}	{{.FrequencyType}}	{{.RetentionValue}} {{.RetentionUnit}}
+{{- end}}
+{{if .OnDemandPolicyItem}}{{.OnDemandPolicyItem.Id}}	-	{{.OnDemandPolicyItem.FrequencyType}}	{{.OnDemandPolicyItem.RetentionValue}} {{.OnDemandPolicyItem.RetentionUnit}}{{end}}
+`
```
Comment (gssbzn) [inline internal/cli/atlas/backup/compliancepolicy/policies/update/update.go:61]:
```suggestion
var updateWatchTemplate = `Your backup compliance policy has been updated
`
```

File: internal/cli/atlas/backup/compliancepolicy/policies/update/update.go [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+	"frequencyType": "daily",
+	"id": "stringstringstringstring",
+	"retentionUnit": "days",
+	"retentionValue": 0
+}
+
+To get the ID of a policy item, run "atlas backups compliancepolicy policies describe".
+
+Finally, run the command as such: "atlas backups compliancepolicy policies update --file /path/to/file"
+`
```
Comment (gssbzn) [inline internal/cli/atlas/backup/compliancepolicy/policies/update/update.go:76]: see other PR on examples needing a particular format

File: internal/cli/atlas/backup/compliancepolicy/policies/update/update.go (hunk 2 above)
Comment (gssbzn) [reply internal/cli/atlas/backup/compliancepolicy/policies/update/update.go:80]: why is the 500? is the json formatted wrongly? missing fields? incompatible fields?

**Comment by VincentPikand** [comment]
closed due to pin
see #2184

**Pull request closed without merging**